drivers/dma/dma_sedi.c

```diff
@@ -0,0 +1,533 @@
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *local_config = &(data->dma_configs[channel]);
+
+	memcpy(local_config, config, sizeof(struct dma_config));
+
+	local_config->head_block =
+		alloc_and_cpy_blk_nodes(config->head_block);
+
+	if (local_config->head_block == NULL) {
+		return -1;
+	}
+
+	/* initialize the dma controller, following the sedi api*/
+	sedi_dma_event_cb_t cb = dma_handler;
+
+	sedi_dma_init(info->peripheral_id, (int)channel, cb, (void *)dev);
+
+	return 0;
+
+INVALID_ARGS:
+	return -1;
+}
+
+static int dma_sedi_reload(const struct device *dev, uint32_t channel,
+			      uint64_t src, uint64_t dst, size_t size)
+{
+	if ((dev == NULL) || (channel >= DEV_CFG(dev)->chn_num)) {
+		LOG_ERR("dma reload failed for invalid args");
+		return -ENOTSUP;
+	}
+
+	int ret = 0;
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *config = &(data->dma_configs[channel]);
+	struct dma_block_config *block_config;
+
+	if ((config == NULL) || (config->head_block == NULL)) {
+		LOG_ERR("dma reload failed, no config found");
+		return -ENOTSUP;
+	}
+	block_config = config->head_block;
+
+	if ((config->block_count == 1) || (block_config->next_block == NULL)) {
+		block_config->source_address = src;
+		block_config->dest_address = dst;
+		block_config->block_size = size;
+	} else {
+		LOG_ERR("no reload support for multi-linkedlist mode");
+		return -ENOTSUP;
+	}
+	return ret;
+}
+
+static int dma_sedi_start(const struct device *dev, uint32_t channel)
+{
+	if ((dev == NULL) || (channel >= DEV_CFG(dev)->chn_num)) {
+		LOG_ERR("dma transferring failed for invalid args");
+		return -ENOTSUP;
+	}
+
+	int ret = -1;
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *config = &(data->dma_configs[channel]);
+	struct dma_block_config *block_config = config->head_block;
+	uint64_t src_addr, dst_addr;
+
+	if (config->block_count == 1) {
+		/* call sedi start function */
+		ret = dma_sedi_apply_single_config(info->peripheral_id,
+						   channel, config);
+		if (ret) {
+			goto ERR;
+		}
+		src_addr = block_config->source_address;
+		dst_addr = block_config->dest_address;
+
+		ret = sedi_dma_start_transfer(info->peripheral_id, channel,
+						src_addr, dst_addr, block_config->block_size);
+	} else {
+		LOG_ERR("MULTIPLE_BLOCK CONFIG is not set");
+		goto ERR;
+	}
+
+	if (ret != SEDI_DRIVER_OK) {
+		goto ERR;
+	}
+
+	return ret;
+
+ERR:
+	LOG_ERR("dma transfer failed");
+	return ret;
+}
+
+static int dma_sedi_stop(const struct device *dev, uint32_t channel)
+{
+	const struct dma_sedi_config_info *const info = DEV_CFG(dev);
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *config = &(data->dma_configs[channel]);
+
+	LOG_DBG("stopping dma: %p, %d", dev, channel);
+	sedi_dma_abort_transfer(info->peripheral_id, channel);
+
+	free_blk_nodes(config);
```

Comment:
If done the way edma or dw gpdma does this, no free is necessary here, the blocks can simply be reused.

There's no guarantee dma_stop is called, one shot transfers, e.g. dma_config + dma_start are entirely acceptable.

Reply:
Has changed this code part, Thanks